### DISCUSSION TITLE
Span for reserved keywords

### DISCUSSION THREAD

**Discussion by magic-akari** [swc-project/swc#5385] [Q&A]
Is it worth saving span for reserved keywords?

I found this issue when I was investigating the comments of `export default`.

Input

```JavaScript
/*1*/ export /*2*/ default /*3*/ function /*4*/ foo /*5*/() /*6*/ {} /*7*/
```

[SWC](https://play.swc.rs/?version=1.2.222&code=H4sIAAAAAAAAA9PXMtTSV0itKMgvKlHQ1zICclJS0xJLc0A8YyAvrTQvuSQzPw%2FINQFx8%2FOBLFMtfQ1NIG0GFKmuBTLMtfQBr22gEUoAAAA%3D&config=H4sIAAAAAAAAA6tWyipOVrKqVipILCpOLQKxiivzShIrlKyUUpNzE4uTizILSpRqdZRKEovSU0tAwsVGBkZGIKHc%2FJTSnFSQnpLKglSwlJlSbS0AqIGbXVQAAAA%3D)

```JavaScript
/*1*/ export default /*3*/ function /*4*/ foo() /*6*/ {}; /*7*/
```

[Babel](https://babeljs.io/repl/#?browsers=chrome%20latest%0A&build=&builtIns=false&corejs=3.21&spec=false&loose=false&code_lz=PQKgjCwAQKYB4AcD2AnALlUAmSUAmMAZgIYCuANhqAMy6GkB2AxmgJZIOYgAsdSSXAKyQAFAEouANlwBvAL5cA7JCA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=false&presets=&prettier=false&targets=&version=7.18.10&externalPlugins=&assumptions=%7B%7D)

```JavaScript
/*1*/
export default
/*3*/
function
/*4*/
foo
/*5*/
()
/*6*/
{}
/*7*/
```

[tsc](https://www.typescriptlang.org/play?#code/PQKgjCwAQKYB4AcD2AnALlUAmSUAmMAZgIYCuANhqAMy6GkB2AxmgJZIOYgAsdSSXAKyQAFAEouANlwBvAL5cA7JCA)

```JavaScript
/*1*/ export /*2*/ default /*3*/ function foo /*5*/() { } /*7*/
```

[Rome](https://play.rome.tools/?code=%2F*1*%2F+export+%2F*2*%2F+default+%2F*3*%2F+function+%2F*4*%2F+foo+%2F*5*%2F%28%29+%2F*6*%2F+%7B%7D+%2F*7*%2F&lineWidth=80&indentStyle=tab&quoteStyle=double&indentWidth=2&sourceType=module&typescript=false&jsx=false#LwAqADEAKgAvACAAZQB4AHAAbwByAHQAIAAvACoAMgAqAC8AIABkAGUAZgBhAHUAbAB0ACAALwAqADMAKgAvACAAZgB1AG4AYwB0AGkAbwBuACAALwAqADQAKgAvACAAZgBvAG8AIAAvACoANQAqAC8AKAApACAALwAqADYAKgAvACAAewB9ACAALwAqADcAKgAvAA==)

```JavaScript
/*1*/ export /*2*/ default /*3*/ function /*4*/ foo /*5*/ () /*6*/ {} /*7*/
```

What the Rome tool shows is the clearest, and you can see directly that it stores information about the `default` keyword in the AST.

<img width="879" alt="image" src="https://user-images.githubusercontent.com/7829098/182657537-2fb27aa5-a51a-4e2a-a941-43f8e2be309a.png">

If we expect the comments surrounding the keywords to be output, then we would have to store some additional span information.

## Is this worth it?

And I found more unit test cases in tsc repo. For example

```JavaScript
/*a*/ do /*b*/ {} /*c*/ while /*d*/ ( /*e*/ true /*f*/ );
```

The comments around the `while` are lost in SWC output.

---

Note: esbuild always drops comments for performance concern.

**Comment by kdy1** [comment]
Maybe? I considered changing `bool`s to `Option<Span>` at the past but I forgot about it
